janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java

```diff
@@ -4720,13 +4724,88 @@ public void testBatchPropertiesPrefetching(int txCacheSize){
-            assertFalse(traversalMetricsStr.contains("multiPreFetch=true"));
+            assertNotContains(janusGraphVertexStepMetrics, "multiPreFetch", "true");
+        }
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = {0, 1, 2, 3, 10, 15, Integer.MAX_VALUE})
+    public void testBatchPropertiesPrefetchingFromEdges(int txCacheSize){
+        boolean inmemoryBackend = getConfig().get(STORAGE_BACKEND).equals("inmemory");
+        int numV = 10;
+        int expectedVerticesPrefetch = Math.min(txCacheSize, 4);
+        JanusGraphVertex mainVertex = graph.addVertex("id", 0);
+        for (int i = 1; i <= numV; i++) {
+            JanusGraphVertex adjacentVertex = graph.addVertex("id", i);
+            mainVertex.addEdge("knows", adjacentVertex, "id", i);
+        }
+        graph.tx().commit();
+
+        if(!inmemoryBackend){
+            clopen(option(BATCH_PROPERTY_PREFETCHING), true, option(TX_CACHE_SIZE), txCacheSize);
+        }
+        GraphTraversalSource gts = graph.traversal();
+
+        for(Direction direction : Direction.values()){
+            GraphTraversal<Vertex, Edge> graphEdgeTraversal = gts.V().has("id", 0).outE("knows")
+                .has("id", P.within(4,5,6,7));
+            GraphTraversal<Vertex, Vertex> graphVertexTraversal;
+
+            switch (direction){
+                case IN: graphVertexTraversal = graphEdgeTraversal.inV(); break;
+                case OUT: graphVertexTraversal = graphEdgeTraversal.outV(); break;
+                case BOTH: graphVertexTraversal = graphEdgeTraversal.bothV(); break;
+                default: throw new NotImplementedException("No implementation found for direction: "+direction.name());
+            }
+
+            TraversalMetrics traversalMetrics = graphVertexTraversal
+                .has("id", P.within(4,5,6,7)).values("id").profile().next();
+
+            Metrics janusGraphEdgeVertexStepMetrics = getStepMetrics(traversalMetrics, JanusGraphEdgeVertexStep.class);
+            if(inmemoryBackend){
+                assertNull(janusGraphEdgeVertexStepMetrics);
+            } else {
+                assertNotNull(janusGraphEdgeVertexStepMetrics);
+                assertTrue(janusGraphEdgeVertexStepMetrics.getName().endsWith("("+direction.name()+")"));
+                if(expectedVerticesPrefetch>1 && !OUT.equals(direction)){
+                    assertContains(janusGraphEdgeVertexStepMetrics, "multiPreFetch", "true");
+                    boolean withAdditionalOutVertex = BOTH.equals(direction) && txCacheSize > 4; // 4 is the number or retrieved IN vertices
+                    assertContains(janusGraphEdgeVertexStepMetrics, "vertices", expectedVerticesPrefetch + (withAdditionalOutVertex?1:0));
+                } else {
+                    assertNotContains(janusGraphEdgeVertexStepMetrics, "multiPreFetch", "true");
+                }
+            }
+        }
+    }
+
+    private void assertContains(Metrics metrics, String annotationKey, Object annotationValue){
```

Comment:
(nitpick) you may consider moving these helper methods to `JanusGraphAssert.java`.

Reply:
I was actually looking for this class. Forgot it's name :smile: Thank you! Fixed